sdk/plugin/v5/plugin.go

```diff
@@ -0,0 +1,180 @@
+type BackendPluginClient struct {
+	client pluginutil.PluginClient
+	sync.Mutex
+
+	logical.Backend
+}
+
+type ContextKey string
+
+func (c ContextKey) String() string {
+	return "plugin" + string(c)
+}
+
+var ContextKeyPluginReload = ContextKey("plugin-reload")
+
+// Cleanup cleans up the go-plugin client and the plugin catalog
+func (b *BackendPluginClient) Cleanup(ctx context.Context) {
+	_, ok := ctx.Value(ContextKeyPluginReload).(string)
+	if !ok {
+		b.Backend.Cleanup(ctx)
+		b.client.Close()
+		return
+	}
+	b.Backend.Cleanup(ctx)
+	b.client.Reload()
+}
+
+// NewBackend will return an instance of an RPC-based client implementation of the backend for
+// external plugins, or a concrete implementation of the backend if it is a builtin backend.
+// The backend is returned as a logical.Backend interface. The isMetadataMode param determines whether
+// the plugin should run in metadata mode.
+func NewBackend(ctx context.Context, pluginName string, pluginType consts.PluginType, sys pluginutil.LookRunnerUtil, conf *logical.BackendConfig, isMetadataMode bool) (logical.Backend, error) {
+	// Look for plugin in the plugin catalog
+	pluginRunner, err := sys.LookupPlugin(ctx, pluginName, pluginType)
+	if err != nil {
+		return nil, err
+	}
+
+	var backend logical.Backend
+	if pluginRunner.Builtin {
+		// Plugin is builtin so we can retrieve an instance of the interface
+		// from the pluginRunner. Then cast it to logical.Factory.
+		rawFactory, err := pluginRunner.BuiltinFactory()
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting plugin type: {{err}}", err)
+		}
+
+		if factory, ok := rawFactory.(logical.Factory); !ok {
+			return nil, fmt.Errorf("unsupported backend type: %q", pluginName)
+		} else {
+			if backend, err = factory(ctx, conf); err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// create a backendPluginClient instance
+		config := pluginutil.PluginClientConfig{
+			Name:            pluginName,
+			PluginSets:      PluginSet,
+			PluginType:      pluginType,
+			HandshakeConfig: bplugin.HandshakeConfig,
+			Logger:          conf.Logger.Named(pluginName),
+			IsMetadataMode:  isMetadataMode,
+			AutoMTLS:        true,
+			Wrapper:         sys,
+		}
+		backend, err = NewPluginClient(ctx, sys, config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return backend, nil
+}
+
+// PluginSet is the map of plugins we can dispense.
+var PluginSet = map[int]plugin.PluginSet{
+	// Version 3 used to supports both protocols. We want to keep it around
+	// since it's possible old plugins built against this version will still
+	// work with gRPC. There is currently no difference between version 3
+	// and version 4.
+	3: {
+		"backend": &bplugin.GRPCBackendPlugin{},
+	},
+	4: {
+		"backend": &bplugin.GRPCBackendPlugin{},
+	},
+	5: {
+		"backend": &bplugin.GRPCBackendPlugin{},
+	},
+}
+
+func Dispense(ctx context.Context, rpcClient plugin.ClientProtocol, pluginClient pluginutil.PluginClient) (logical.Backend, error) {
```

Comment:
Nit: Unused context here. You may also want to keep it around for future use (exported SDK method), so feel free to disregard if so.